File: rust/signal-server/src/sigserve.rs
use std::ffi::{c_char, c_int, CString, NulError};
use thiserror::Error;

#[derive(Error, Debug)]
pub enum SigserveError {
    #[error("arg contains a null byte: {0}")]
    Args(#[from] NulError),
    #[error("libsigserve returned a non-zero integer {0}")]
    Retcode(c_int),
}

pub fn call_sigserve(args: &str) -> Result<(), SigserveError> {
    let cstrings = args
        .split_whitespace()
        .map(CString::new)
        .collect::<Result<Vec<CString>, _>>()?;
    let mut c_args = cstrings
        .iter()
        .map(|arg| arg.as_ptr() as *mut c_char)
        .collect::<Vec<*mut c_char>>();
    // SAFETY: we're calling into a >20 y/o C+ codebase. Safety can
    //         not be guaranteed.
    match unsafe { ffi::entry_point(c_args.len() as i32, c_args.as_mut_ptr()) } {
        0 => Ok(()),
        other => Err(SigserveError::Retcode(other)),
    }
}
#[cxx::bridge(namespace = "sigserve_wrapper")]
mod ffi {
    unsafe extern "C++" {
        include!("signal-server/src/sigserve.h");
        unsafe fn entry_point(argc: i32, argv: *mut *mut c_char) -> i32;
    }
}

File: rust/signal-server/src/sigserve.cc
#include "sigserve.h"

#include "../../../src/common.hh"
#include "signal-server/src/sigserve.rs.h"

extern int init(const char *sdf_path, bool debug);
extern int handle_args(int argc, char *argv[], output *ret_out);

namespace sigserve_wrapper {

int init(const char *sdf_path, bool debug) { return ::init(sdf_path, debug); };
Report handle_args(int argc, char *argv[])
{
    Report report;
    report.dbm = 0.0;
    report.loss = 0.0;
    report.field_strength = 0.0;

    output out;

    report.retcode = ::handle_args(argc, argv, &out);
    if (report.retcode) {
        return report;
    }

    report.dbm = out.dBm;
    report.loss = out.loss;
    report.field_strength = out.field_strength;

    for (auto &a : out.cluttervec) {
        report.cluttervec.push_back(a);
    }

    for (auto &a : out.referencevec) {
        report.referencevec.push_back(a);
    }

    for (auto &a : out.fresnelvec) {
        report.fresnelvec.push_back(a);
    }

    for (auto &a : out.fresnel60vec) {
        report.fresnel60vec.push_back(a);
    }

    for (auto &a : out.curvaturevec) {
        report.curvaturevec.push_back(a);
    }

    for (auto &a : out.profilevec) {
        report.profilevec.push_back(a);
    }

    for (auto &a : out.distancevec) {
        report.distancevec.push_back(a);
    }


    report.image_data.reserve(out.imagedata.size());
    for (auto &pixel : out.imagedata) {
        report.image_data.push_back(pixel);
    }

    return report;
}

}  // namespace sigserve_wrapper
